File: codechef/long-challenges/2021/jan/wipl.cpp
#include <stdio.h>
#include <stdlib.h>
#include <algorithm>
#include <vector>

using namespace std;

int dp[100000];
int temp_dp[100000];
int previous[100000];

int heights[40001];
int used_boxes[40001];

int main()
{
    int tests;
    int n;
    int k;
    int i;
    int j;
    int l;
    int tracy_height;
    int count;
    int least_count;
    int max_dp;
    int found;
    int cur_height;

    scanf("%d", &tests);
    while (tests--)
    {
        scanf("%d %d", &n, &k);
        for (i = 0; i < n; i++)
        {
            scanf("%d", &heights[i]);
        }

        // sort in descending
        sort(heights, heights + n);
        for (j = 0; j < n/2; j++)
        {
            l = heights[j];
            heights[j] = heights[n - 1 - j];
            heights[n - 1 - j] = l;
        }

        if (n == 1)
        {
            printf("-1\n");
            continue;
        }

        if (heights[1] >= k)
        {
            printf("2\n");
            continue;
        }

        if (heights[0] >= k)
        {
            // find greedy and continue
            tracy_height = 0;
            count = 1;
            j = 1;
            while (tracy_height < k && j < n)
            {
                count++;
                tracy_height += heights[j];
                j++;
            }

            if (tracy_height >= k)
            {
                printf("%d\n", count);
            }
            else
            {
                printf("-1\n");
            }
            
            continue;
        }

        // default to DP case
        max_dp = 2 * k;
        for (i = 0; i < 100000; i++)
        {
            dp[i] = temp_dp[i] = n + 1;
        }

        for (i = 0; i < n; i++)
        {
            dp[heights[i]] = 1;
            previous[heights[i]] = i;

            for (j = heights[i] + 1; j < max_dp; j++)
            {
                if ((temp_dp[j - heights[i]] + 1) < dp[j])
                {
                    dp[j] = temp_dp[j - heights[i]] + 1;
                    previous[j] = i;
                }
            }

            for (j = heights[i]; j < max_dp; j++)
            {
                temp_dp[j] = dp[j];
            }
        }

        // now find a combination
        found = 0;
        least_count = n + 1;

        for (j = k; j < max_dp; j++)
        {
            if (dp[j] < n && dp[j] < least_count)
            {
                for (i = 0; i <= 40000; i++)
                {
                    used_boxes[i] = 0;
                }

                cur_height = j;
                while (cur_height > 0)
                {
                    used_boxes[previous[cur_height]] = 1;
                    cur_height -= heights[previous[cur_height]];
                }

                // now do greedy
                tracy_height = 0;
                count = dp[j];
                i = 0;

                while ((tracy_height < k) && (i < n))
                {
                    if (!used_boxes[i])
                    {
                        tracy_height += heights[i];
                        count++;
                    }

                    i++;
                }

                if (tracy_height >= k)
                {
                    found = 1;

                    if (count < least_count)
                    {
                        least_count = count;
                    }
                }
            }
        }

        if (found)
        {
            printf("%d\n", least_count);
        }
        else
        {
            printf("-1\n");
        }
    }

    return 0;
}

File: codechef/long-challenges/2021/jan/wipl.py
from sys import stdin, stdout

tests = int(stdin.readline())

while tests > 0:
    tests -= 1

    n, k = map(int, stdin.readline().split())
    heights_map = map(int, stdin.readline().split())
    heights = list(heights_map)

    if n == 1:
        print(-1)
        continue

    heights.sort()
    heights.reverse()

    # list has two boxes taller than needed
    if heights[1] >= k:
        print(2)
        continue
    # list has one tall enough. remove it and compute the other
    # obviously, the one tall enough is first one
    elif heights[0] >= k:
        tracy = 0
        count = 1
        # pick that one and build another
        for i in range(1, len(heights)):
            count += 1
            tracy += heights[i]
            if tracy >= k:
                break

        if tracy >= k:
            print(count)
        else:
            print(-1)
        continue

    # no trivial cases present
    # compute DP to find out minimum needed to reach closest to k
    max_dp = k + heights[0]
    dp = [len(heights) + 1] * max_dp
    temp_dp = [len(heights) + 1] * max_dp
    prev = [-1] * (2 * k)

    heights_len = len(heights)

    for j in range(0, heights_len):
        for i in range(heights[j], max_dp):
            if i == heights[j]:
                dp[i] = 1
                prev[i] = j
            else:
                if (temp_dp[i - heights[j]] + 1) < dp[i]:
                    dp[i] = temp_dp[i - heights[j]] + 1
                    prev[i] = j
        for i in range(1, max_dp):
            temp_dp[i] = dp[i]

    least_count = (heights_len + 1) * 2.0
    count_found = False
    # find the combination with least no. of boxes
    for h in range(k, max_dp):
        if dp[h] < heights_len and dp[h] < (least_count / 2.0):
            # there's a potential combination
            # compute map with used up indices
            cur_height = h
            used_boxes = []
            cur_prev = prev[h]
            while cur_height > 0:
                used_boxes.append(cur_prev)
                cur_height -= heights[cur_prev]
                cur_prev = prev[cur_height]

            # now start in descending order and compute tracy's height
            tracy = 0
            count = 0
            box_idx = 0
            while tracy < k and box_idx < heights_len:
                if used_boxes.__contains__(box_idx):
                    box_idx += 1
                else:
                    count += 1
                    tracy += heights[box_idx]
                    box_idx += 1

            if tracy >= k and least_count > (dp[h] + count):
                count_found = True
                least_count = dp[h] + count
        if count_found:
            break

    if count_found:
        print(least_count)
    else:
        print(-1)
